### PR TITLE
issue #27642 - timedelta merge asof with tolerance

### LINKED ISSUE #27642: merge_asof(): cannot use tolerance flag when the index is a TimedeltaIndex
#### Code Sample

```python
import pandas as pd 
import numpy as np

print(
    """
    \nPandas merge_asof() bug:
    
    \tUnimplemented error?
    \tcannot use tolerance flag when my index is a timedelta (not a timestamp)
    \tjust documenting so I can try to add this functionality
        
    """)

print(f"pandas version: {pd.__version__}")
print(f"numpy version: {np.__version__}")

delta_300 = pd.timedelta_range(start='0 minutes', freq='3333334 N', periods=301, name='Time')
delta_120 = pd.timedelta_range(start='0 minutes', freq='8333334 N', periods=301, name='Time')

df_300hz = pd.DataFrame({'my300hz_data': np.arange(301)}, index=delta_300)
df_120hz = pd.DataFrame({'my120hz_data': np.arange(301)}, index=delta_120)

print(df_300hz)
print(df_120hz) 

### this throws error "pandas.errors.MergeError: key must be integer, timestamp or float"
merged = pd.merge_asof(df_120hz, df_300hz, on='Time', direction='nearest', tolerance=pd.Timedelta('15 ms'))

### The line below works, but output is not what I want
# merged = pd.merge_asof(df_120hz, df_300hz, on='Time', direction='nearest')

merged.set_index('Time', inplace=True)
print(merged)
```

#### Problem description

I need to see NaNs when I merge and there is a gap in my data, without begin able to use the tolerance flag my data gets smoothed.

Current work around, convert all my TimedeltaIndex's to a time stamp. Since I do not have a date for this data I am using unix time. This feels bulky since I am going to drop the date when I save the file anyway.

#### Expected Output

[301 rows x 1 columns]
                 my120hz_data  my300hz_data
Time                                       
00:00:00                    0             0
00:00:00.008333             1             2
00:00:00.016666             2             5
00:00:00.025000             3             7
00:00:00.033333             4            10
...                       ...           ...
00:00:02.466666           296           NaN
00:00:02.475000           297           NaN
00:00:02.483333           298           NaN
00:00:02.491666           299           NaN
00:00:02.500000           300           NaN

#### Output of ``pd.show_versions()``

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.3.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.19.0-5-amd64
machine          : x86_64
processor        : 
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 0.25.0
numpy            : 1.17.0
pytz             : 2019.1
dateutil         : 2.8.0
pip              : 18.1
setuptools       : 40.8.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
s3fs             : None
scipy            : 1.3.0
sqlalchemy       : None
tables           : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None

### PR DESCRIPTION
**Pull request by ianzur**
- [X] closes #27642 
- [x] tests 2 / 1 [added / passed] 
- [X] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry
- [ ] \(Optional) allow users to pass datetime.timedelta objects for tolerance

### CODE REVIEW

**Comment by ianzur** [comment]
My first PR, please be gentle.

File: pandas/core/reshape/merge.py [hunk 1]
```diff
@@ -1635,7 +1636,11 @@ def _get_merge_keys(self):
                 )
             )
 
-            if is_datetime64_dtype(lt) or is_datetime64tz_dtype(lt):
+            if(
```
Comment (TomAugspurger) [inline pandas/core/reshape/merge.py]: This may throw a linter warning (space after the `if`).

**Review by TomAugspurger — COMMENTED** [review]
Can you add a release note to `doc/source/whatsnew/v0.25.1.rst`?

We'll also want a test ensuring this continues to work in the future. Can you simplify your example from https://github.com/pandas-dev/pandas/issues/27642 and make it a unit test? Make sure you include a reference to the GitHub issue in a comment in the test.

File: pandas/core/reshape/merge.py [hunk 2]
```diff
@@ -1635,7 +1636,11 @@ def _get_merge_keys(self):
...
+                is_datetime64_dtype(lt)
+                or is_datetime64tz_dtype(lt)
+                or is_timedelta64_dtype(lt)
+            ):
                 if not isinstance(self.tolerance, Timedelta):
```
Comment (mroeschke) [inline pandas/core/reshape/merge.py]: This should probably be `if not isinstance(self.tolerance, datetime.timedelta)` so users can pass `datetime.timedelta` objects. If you could add a test case for that situation that'd be great.

Comment (ianzur) [reply pandas/core/reshape/merge.py]: see commit "mroeschke: this test is for you :rocket:"

Comment (ianzur) [reply pandas/core/reshape/merge.py]: the test fails

Comment (ianzur) [reply pandas/core/reshape/merge.py]: We'll need a different PR for this test I think, it fails at this conversion check. 
https://github.com/pandas-dev/pandas/blob/143bc34aa8f068b18e7137df7ca91b9929cc1389/pandas/core/reshape/merge.py#L1709

datetime.timedelta has no attribute "value"

Comment (ianzur) [reply pandas/core/reshape/merge.py]: `datetime.timedelta` objects do not have nanoseconds like `pd.Timedelta` 
Another check and conversion is necessary to convert `datetime.timedelta` correctly to nanoseconds

Comment (mroeschke) [reply pandas/core/reshape/merge.py]: Okay thanks for investigating. Yes, I'd be great to address in another issues/PR.

File: doc/source/whatsnew/v0.25.1.rst [hunk 3]
```diff
@@ -129,7 +129,7 @@ Reshaping
 ^^^^^^^^^
 
 - A ``KeyError`` is now raised if ``.unstack()`` is called on a :class:`Series` or :class:`DataFrame` with a flat :class:`Index` passing a name which is not the correct one (:issue:`18303`)
--
+- Bug in ``.merge_asof(tolerance=tolerance)`` could not merge :class:`Timedelta` objects (:issue:`27642`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.1.rst]: use ``:meth:`merge_asof` ``, and you can mention when passing the ``tolerance`` kwarg

File: pandas/tests/reshape/merge/test_merge_asof.py [hunk 4]
```diff
@@ -599,6 +600,19 @@ def test_tolerance(self):
         expected = self.tolerance
         assert_frame_equal(result, expected)
 
+    def test_datetime_timedelta_tolerance(self):
```
Comment (jreback) [inline pandas/tests/reshape/merge/test_merge_asof.py]: can you parameterize on datetime.timedelta and Timedelta

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/reshape/merge/test_merge_asof.py (hunk 4 above)
Comment (ianzur) [reply pandas/tests/reshape/merge/test_merge_asof.py]: Please take a look, I think I set it up right.

Comment (ianzur) [reply pandas/tests/reshape/merge/test_merge_asof.py]: Have a look, I think I set it up correctly

File: pandas/core/reshape/merge.py (hunk 1 above)
Comment (ianzur) [reply pandas/core/reshape/merge.py]: black wants to put a space here

**Comment by ianzur** [comment]
Does anyone understand why my test fails in python3.5?

**Comment by WillAyd** [comment]
On phone so haven’t checked logs but if 3.5 only most likely the construction of data frames from Dict don’t have a defined order. 

Probably easiest to construct from lists

**Comment by ianzur** [comment]
Do I close the PR?

**Comment by TomAugspurger** [comment]
@ianzur you may want to set up a 3.5 env locally for testing that. `conda env create -n pandas-azure-35-compat -f ci/deps/azure-35-compat.yaml`

**Review by TomAugspurger — APPROVED** [review]

File: pandas/tests/reshape/merge/test_merge_asof.py [hunk 5]
```diff
@@ -588,14 +590,23 @@ def test_non_sorted(self):
...
         merge_asof(trades, self.quotes, on="time", by="ticker")
 
-    def test_tolerance(self):
+    @pytest.mark.parametrize(
+        "tolerance",
+        [
+            Timedelta("1day"),
+            pytest.param(
+                datetime.timedelta(days=1),
+                marks=pytest.mark.xfail(reason="not implemented", strict=True),
```
Comment (TomAugspurger) [inline pandas/tests/reshape/merge/test_merge_asof.py]: Do we have an open issue to implement this?

Comment (ianzur) [reply pandas/tests/reshape/merge/test_merge_asof.py]: No, not that I know of.

**Comment by TomAugspurger** [comment]
Could you open one with an example of the failure right now?

**Comment by ianzur** [comment]
I can later today, I'll @you.

**Pull request merged**